tools/tsec.bzl

```
@@ -0,0 +1,143 @@
+    attrs = {
+        "src": attr.label(
+            mandatory = True,
+            allow_single_file = [".json"],
+            doc = """The single tsconfig used for running tsec.""",
+        ),
+        "deps": attr.label_list(
+            allow_files = [".json"],
+            doc = """Any configs extended by `src`.""",
+        ),
+        "exemption": attr.label(
+            allow_single_file = [".json"],
+            doc = """The exemption list used by `src`.""",
+        ),
+    },
+    doc = """Compute all transitive dependencies of a tsec_test config. """,
+)
+
+TsLibInfo = provider(fields = ["srcs", "deps"])
+
+def _ts_library_forwarded_impl(ctx):
+    """Forward `srcs` and `deps` of `ts_library` and `ng_module` macros to `_tsec_test`."""
+    return [TsLibInfo(srcs = ctx.attr.srcs, deps = ctx.attr.deps)]
+
+ts_library_forwarded = rule(
+    implementation = _ts_library_forwarded_impl,
+    attrs = {
+        "srcs": attr.label_list(allow_files = [".ts", ".tsx"]),
+        "deps": attr.label_list(),
+    },
+    doc = """A rule-in-the-middle to forward `srcs` and `deps` to _all_data, to avoid repeating these arguments in tsec_test.
+
+Should only be used by the `ts_library` and `ng_module` macros in tools/default.bzel.""",
+)
+
+def get_forwarded_target_name(name):
+    return name + "_forwarded"
+
+def _all_data_impl(ctx):
+    files = []
+
+    if TsecTsconfigInfo not in ctx.attr.tsconfig:
+        fail("`tsconfig` must be a tsec_config target")
+
+    tsec_tsconfig_info = ctx.attr.tsconfig[TsecTsconfigInfo]
+    files.append(tsec_tsconfig_info.src)
+    if tsec_tsconfig_info.exemption:
+        files.append(tsec_tsconfig_info.exemption)
+    files.extend(tsec_tsconfig_info.deps)
+
+    if TsLibInfo not in ctx.attr.forwarded_ts_target:
+        fail("`target` must be a ts_library_forwarded target")
+
+    ts_target_info = ctx.attr.forwarded_ts_target[TsLibInfo]
+    for s in ts_target_info.srcs:
+        if hasattr(s, "files"):
+            files.extend(s.files.to_list())
+
+    for d in ts_target_info.deps:
+        if DeclarationInfo in d:
+            files.extend(d[DeclarationInfo].transitive_declarations.to_list())
+        if hasattr(d, "files"):
+            files.extend(d.files.to_list())
+
+    return [DefaultInfo(files = depset(files))]
+
+_all_data = rule(
+    implementation = _all_data_impl,
+    attrs = {
+        "tsconfig": attr.label(),
+        "forwarded_ts_target": attr.label(),
+    },
+    doc = """Expand all transitive dependencies needed to run `_tsec_test`.""",
+)
+
+def tsec_test(name, target, tsconfig):
+    """Run tsec over a ts_library or ng_module target to check its compatibility with Trusted Types.
+
+    This rule DOES NOT check transitive dependencies.
+
+    Args:
+        name: name of the tsec test
+        target: the ts_library or ng_module target to be checked
+        tsconfig: the tsec_config target used for configuring tsec
+    """
+    all_data = name + "_all_data"
```

Historically we have preferred to use python's string formatting for substitutions like this.

```suggestion
    all_data = "%s_all_data" % name
```

---

Can we just do the substitutions in place, or change the variable name used?  It seems odd to have `all_data` as a string that we use as a parameter on `_all_data`

---

Done.